docs/enhancements/auto-return-agent-to-infra-env.md

```diff
@@ -0,0 +1,140 @@
+---
+title: auto-return-agent-to-infra-env
+authors:
+  - "@carbonin"
+creation-date: 2022-05-17
+last-updated: 2022-05-17
+---
+
+# Automatically Return Agents to InfraEnv
+
+## Summary
+
+Without a BareMetalHost there is currently no way to automate the process
+of unbinding a host from a cluster. When a user wants to reuse a host for a new
+cluster they have to use the assisted-service unbind API, then manually boot
+the host back into the discovery image.
+
+To support automating this procedure, the assisted-service will configure nodes
+in installed clusters to run an agent. When appropriate this agent will
+reconfigure the host to boot into a discovery image, allowing the host to
+be made available for installation without the user manually interacting with
+a management system to reboot the host into the discovery image.
+
+## Motivation
+
+A user creating and managing clusters using the assisted-service may not have
+access to the systems used to configure and boot hosts. This is especially
+true in bare-metal use cases where a BareMetalHost is not present. The
+assisted-service can make this use case easier by handling the reprovisioning
+process.
+
+### Goals
+
+* Reconfigure a host to boot the discovery image as a part of unbinding from a
+  cluster without a BareMetalHost
+* Allow for a node to be gracefully removed from an installed cluster
+
+### Non-Goals
+
+* Automate graceful removal of a node from an existing cluster
+* Booting a node that is unresponsive or not currently part of an installed
+  cluster
+
+## Proposal
+
+When a host should be returned to the InfraEnv use the kubeconfig created when
+the cluster was installed to run an agent on the host. This agent will handle
+downloading the required artifacts and rebooting the host into the discovery
+process.
+
+### Timeline
+
+1. User triggers an action which will prepare a host to be unbound
+  * Assisted service will create a MachineConfig to run our agent on the host as
+    a systemd service.
+  * The agent will download the InfraEnv's kernel and initrd (including
+    discovery ignition) to the host's `/boot` partition.
+  * The agent will create a GRUB boot entry to boot into the discovery kernel
+    and initrd.
+2. User (or other automation) removes the node from the cluster
+  * Drain existing workloads and delete the node
+3. User unbinds the host from the cluster using the assisted-service API
+  * This triggers the agent to reboot the host into the discovery image
+4. Host is registered as `known-unbound` in the InfraEnv and is ready to be used
+   in new clusters
+
+### User Stories
+
+#### Story 1
```

Comment:
lets think about hypershift/capi as well here
@eranco74 @avishayt

Reply:
Right, this is the assisted-specific portion of that process.
Presumably hypershift or CAPI would utilize this new capability in assisted to make the machine removal process better.

Reply:
our main assumption, iiuc, is that the removal itself is done out of the assisted installer, and the automated part is triggering the unbounding and reuse of the host in another provisioning.

Reply:
i don't think that we will be able to run the agent before the host is unbound, if i remember correctly once it will try to  register the sate machine will reject it because the host is installed, we will probably need to change that, or change the order.

Reply:
So currently, what happens is (1 & 4).
Once CAPI sees that a node should be removed from a cluster it removes it from the cluster. We need to figure out how to block  CAPI from removing the node until 2 & 3 happen.

Reply:
The CAPI provider also does 5 (unbind)

Reply:
>if i remember correctly once it will try to register the sate machine will reject it because the host is installed, we will probably need to change that, or change the order.

Yeah, I think the state machine will need some changes as a part of this (we talked about this today).
Perhaps the agent will need to query the service before attempting to register to allow the service to know the difference between an agent running on an installed node and one running as a part of discovery.

Reply:
>Once CAPI sees that a node should be removed from a cluster it removes it from the cluster. We need to figure out how to block CAPI from removing the node until 2 & 3 happen.

>The CAPI provider also does 5 (unbind)

It sounds like for this to work we will need to break some of this up.
If CAPI is going to trigger us to move a host back into the infraenv using this process, then the host needs to still be a part of the cluster in order for us to get the agent running on it.

At the very least we need CAPI to unbind(or whatever this new process gets called) before it removes the node from the cluster.

Reply:
If/when capi or mapi is involved, I think it makes more sense for it to start the agent on the node being removed. The mapi/capi provider is guaranteed to always have access to the k8s API.

Reply:
As far as I know CAPI is the main use case here, right? @avishayt @filanov ?

The only problem with that is that the CAPI provider wouldn't (and shouldn't IMO) know how to correctly start our agent for a particular infraenv. We could change this proposal to expose an endpoint that would return the ignition that we would put in a machine config for a specific host though.

Reply:
I would think a MachineSet would be associated with an InfraEnv.

Reply:
> I would think a MachineSet would be associated with an InfraEnv.

I don't think this is the case currently, but I also don't think it matters.
As far as I can tell the CAPI provider just maps AgentMachines to Agents and has no knowledge of the infraEnv.
That said it could presumably find the infraEnv based on the label on the agent.

Reply:
Today's implementation:
1. CAPI cordons the node and then deletes the AgentMachine.
2. The CAPI provider reconciles when the AgentMachine is being deleted. It has a finalizer where it unbinds the agent and removes the finalizer.

Proposed implementation:
1. Upon creating an AgentMachine, the CAPI provider will also add a [lifecycle hook](https://github.com/kubernetes-sigs/cluster-api/blob/1ba14a07c47b87a7f2e3bef00cd592e96bb9e8e4/internal/controllers/machine/machine_controller.go#L305) so that it has a chance to boot the agent.  This means that the CAPI provider will need to watch Machines.
2. When the CAPI provider sees that a Machine is waiting on the PreDrainDeleteHook, it will unbind the agent, causing it to reboot, and remove the PreDrainDeleteHook.
3. There is no need for the AgentMachine finalizer.

Reply:
Opened https://issues.redhat.com/browse/MGMT-10968 to handle the CAPI provider changes.

Reply:
Does this seem resolved to all of you?
Don't want to close it out if there are still questions.